cpp/src/lists/sequences.cu

```diff
@@ -0,0 +1,237 @@
+    return starts.element<T>(list_idx) + multiply(steps.element<T>(list_idx), idx - list_offset);
+  }
+};
+
+template <typename T>
+struct tabulator_fixed_step {
+  column_device_view const starts;
+  offset_type const* const offsets;
+  size_type const* const labels;
+
+  auto __device__ operator()(size_type idx) const
+  {
+    auto const list_idx    = labels[idx] - 1;  // labels are 1-based indices
+    auto const list_offset = offsets[list_idx];
+    return starts.element<T>(list_idx) + static_cast<T>(idx - list_offset);
+  }
+};
+
+template <typename T, typename Enable = void>
+struct sequences_functor {
+  template <typename... Args>
+  static std::unique_ptr<column> invoke(Args&&...)
+  {
+    CUDF_FAIL("Unsupported per-list sequence type-agg combination.");
+  }
+};
+
+struct sequences_dispatcher {
+  template <typename T>
+  std::unique_ptr<column> operator()(size_type n_elements,
+                                     column_view const& starts,
+                                     std::optional<column_view> const& steps,
+                                     offset_type const* offsets,
+                                     size_type const* labels,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    return sequences_functor<T>::invoke(n_elements, starts, steps, offsets, labels, stream, mr);
+  }
+};
+
+template <typename T>
+static constexpr bool is_supported()
+{
+  return (cudf::is_numeric<T>() && !cudf::is_boolean<T>()) || cudf::is_duration<T>();
+}
+
+template <typename T>
+struct sequences_functor<T, std::enable_if_t<is_supported<T>()>> {
+  static std::unique_ptr<column> invoke(size_type n_elements,
+                                        column_view const& starts,
+                                        std::optional<column_view> const& steps,
+                                        offset_type const* offsets,
+                                        size_type const* labels,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+  {
+    auto result =
+      make_fixed_width_column(starts.type(), n_elements, mask_state::UNALLOCATED, stream, mr);
+    if (starts.is_empty()) { return result; }
+
+    auto const result_begin  = result->mutable_view().template begin<T>();
+    auto const starts_dv_ptr = column_device_view::create(starts, stream);
+    auto const steps_dv_ptr  = steps ? column_device_view::create(steps.value(), stream) : nullptr;
+
+    if (steps) {
+      auto const op = tabulator<T>{*starts_dv_ptr, *steps_dv_ptr, offsets, labels};
+      thrust::tabulate(rmm::exec_policy(stream), result_begin, result_begin + n_elements, op);
+    } else {
+      auto const op = tabulator_fixed_step<T>{*starts_dv_ptr, offsets, labels};
+      thrust::tabulate(rmm::exec_policy(stream), result_begin, result_begin + n_elements, op);
+    }
```

Comment:
I think this all could be simplified quite a bit by instead of using a `std::optional` for the `steps`, making the `detail` version of `sequences` an iterator based algorithm where you can just pass a constant iterator of `1` for the single step overload. Then you wouldn't need separate dispatches here.

Reply:
Passing an iterator from the beginning is not possible as we need to know type of that iterator thus it must be generated in a dispatched function. 

I have simplified my implementation, removing the `tabulator_fixed_step` struct, using only device data pointer instead of `column_device_view` and set the `steps` pointer to nullptr if `steps` is not available.